### DISCUSSION TITLE
Consistent CLI generation of entities across all database backends

### DISCUSSION THREAD

**Discussion by kriswuollett** [SeaQL/sea-orm#1825] [General]
## Motivation

Stop needing to manually fix entities generated from the database because mapping is ambiguous, e.g., #924, #375, etc. For me it was a `DateTime<Utc>` getting mapped to a `String` in the entity because I was generating from a migrated `sqlite` database.

I don't believe it is possible to generate entities properly from a database instance, i.e., a physical database model, since the types may be erased and resolved to a single type like string in sqlite.

## Proposed Solutions

- serialize migration commands so that the complete logical model is known
  - into the migrations table
  - or, a file that the entity generate command uses
  - or, have the entity generate command make use of the migration library directly

## Current Workarounds

None that I'm aware of. I don't believe reusing database comments as in https://github.com/SeaQL/sea-orm/issues/924#issuecomment-1200852430 would be acceptable since database comments for things like columns can be set in the `ColumnSpec`.

## Related

Defining more custom types like in [entity/prelude.rs](https://github.com/SeaQL/sea-orm/blob/a5f2ae69e9f08648c83e951a2b1ad298bf134d04/src/entity/prelude.rs)?

**Comment by tyt2y3** [comment]
Thank you for you suggestion. There are some low hanging fruits we can do to achieve schema reproducibility for SQLite.
In particular, we can 'play around' SQLite's type affinity to represent the logical types.
If you'd like to help, an investigation on how we can map these types in SeaQuery would be appreciated.
The next step would then be a reverse map in SeaSchema.

Just a side note: when I am working with SQLite I typically won't use migration. I'd `create_table_from_entity`.

**Comment by tyt2y3** [comment]
> serialize migration commands so that the complete logical model is known

It already works this way.

Under the hood, the output of SeaSchema discovery is a SeaQuery SchemaStatement, which will then be feed into the entity generator.
It is possible today to feed a handwritten SeaQuery Statement into `sea-orm-codegen`.
It's just there is no easy path through the cli.

**Comment by kriswuollett** [comment]
> Under the hood, the output of SeaSchema discovery is a SeaQuery SchemaStatement, which will then be feed into the entity generator. It is possible today to feed a handwritten SeaQuery Statement into `sea-orm-codegen`. It's just there is no easy path through the cli.

So basically it should be possible to have a [`build.rs` like I'd have for `tonic`](https://github.com/hyperium/tonic/blob/62a731b95b0a7cacf7a3016412b10e2d9484bd29/examples/build.rs), by 
feeding the [statements into the `EntityTransformer` like in the cli](https://github.com/SeaQL/sea-orm/blob/af52e86a9ced16156899f682011713363e071fcd/sea-orm-cli/src/commands/generate.rs#L185). And that one could look at probably [`exec_up`](https://github.com/SeaQL/sea-orm/blob/af52e86a9ced16156899f682011713363e071fcd/sea-orm-migration/src/migrator.rs#L358) for related info on how the migrations are run.

But I'm guessing that from mentioning handwritten statements, that means it can't currently be automatic, i.e., there is no reducer / operational transform of the migrations so one could have just a final resolved "logical schema" to feed into the entity generator?

**Reply by tyt2y3** [reply]
> So basically it should be possible to have a [build.rs like I'd have for tonic](https://github.com/hyperium/tonic/blob/62a731b95b0a7cacf7a3016412b10e2d9484bd29/examples/build.rs), by
feeding the [statements into the EntityTransformer like in the cli](https://github.com/SeaQL/sea-orm/blob/af52e86a9ced16156899f682011713363e071fcd/sea-orm-cli/src/commands/generate.rs#L185). And that one could look at probably [exec_up](https://github.com/SeaQL/sea-orm/blob/af52e86a9ced16156899f682011713363e071fcd/sea-orm-migration/src/migrator.rs#L358) for related info on how the migrations are run.

Yes. Such a workflow is entirely possible and doable

> so one could have just a final resolved "logical schema" to feed into the entity generator?

The `sea-orm-migration` crate should be able to import `sea-orm-codegen` with no circular dependency. So it's possible for a migration script itself to generate entity files.

What I mean is, the interface between migration and codegen is the in-memory SeaQuery AST, not the textual representation of the SeaQuery code. So for example, if `sea-orm-cli` wants to read a migration file and extract it's SeaQuery Statement, it actually have to compile that file and run it. Alternatively, may be we can make SeaQuery Statements serializable, so that you can create some migration scripts, dump those info into a file, and have `sea-orm-cli` deserialize that dump. May be this is what you really wanted?

**Reply by kriswuollett** [reply]
In short, I'd mainly want to be able to keep code-generated entity structs up to date based without having to use a live database, so I'd have guessed from a calculation on the migration statements. I'm evaluating `sea-orm` on a new service and looking to see what will happen to maintenance as migrations are added.

Basically, I think what I would have expected is `fn migration(Vec<SeaQuery>) -> SeaSchema`, then `fn generate_entities(SeaSchema, SeaEntitiesSchemaMapping, Path)`, but perhaps offline / no live database isn't really in the plans for the project.

**Reply by tyt2y3** [reply]
Umm, I guess we have an understanding here. In short, your use case is entirely possible now, it might require some exploration on the smoothest workflow that suits you.

To clarify again:

`SeaQuery` is an AST representation shared among all components.

The schema statement you usually write is a `SeaQuery` statement.

The migration system:
1. Serialize the SeaQuery statements into SQL
2. Execute the SQL against a live database

The codegen system:
1. Discover the database schema from a live database into a raw `SeaSchema` format
2. Convert the raw `SeaSchema` format into a `SeaQuery` statement
3. Transform the `SeaQuery` statements into an internal Entity representation
https://github.com/SeaQL/sea-orm/blob/af52e86a9ced16156899f682011713363e071fcd/sea-orm-codegen/src/entity/transformer.rs#L12
4. Write the Entities into files

So you can of course jump from migration.1 into codegen.3. You can write a small cli utility to do that. I imagine it can be called something like "bootstrap"?

`schema` <- your source of truth lib with SeaQuery statements
`bootstrap` <- depends on `schema` and generate entities
`migration` <- depends on `schema`, but run migrations against a live database

`entity` <- generated entities
`app` <- you app logic, depending on `entity`

**Reply by tyt2y3** [reply]
In short, your `boostrap` utility only need to do the very last bit of cli: https://github.com/SeaQL/sea-orm/blob/af52e86a9ced16156899f682011713363e071fcd/sea-orm-cli/src/commands/generate.rs#L185-L206

**Comment by kriswuollett** [comment]
So I get that it is possible, but want to call attention to the consistency / live database aspect, since I decided to just use `sea-orm` for its migrator, along with `sea-query` since it gives me more flexibility, and won't use `sea-orm` for the entities for now. Not going to try to write a `reduce(Vec[DdlQuery]) -> Schema` at the moment.

It took me some digging in the code to figure out how to just use the migrations, but I am now running unit tests with a setup like this:

```rust
let temp_dir = tempfile::tempdir().expect("temporary directory");
let db_path = temp_dir.path().to_owned().join("test.sqlite");
let database_url = format!(
    "sqlite:{}?mode=rwc",
     db_path.into_os_string().into_string().expect("path string")
);
let pool = SqlitePoolOptions::new()
    .connect(&database_url)
    .await
    .expect("database pool");
{
    let conn = SqlxSqliteConnector::from_sqlx_sqlite_pool(pool.to_owned());
    Migrator::up(&conn, None).await.expect("database migrated");
}
```

The original issue I filed, #1823, is a cross between a feature request and a bug report. From a practical perspective it seems like the entity generator only works if the schema retrieved from a live database can unambiguously determine the types needed for the entity models. I don't think that is possible especially if things like a mix of different time types are desired. But the scope of everything on my mind, is it would be really interesting if:

- a standalone `sea-migrations` crate exists for those that don't want to use `sea-orm` for the value structs and associated functions
- entities are generated from a model deterministically generated from reading through the `sea-query` DDL statement structs, which I'd guess result in structs found in `sea-schema`, and can be done in a method similar to `tonic_build` in a `build.rs` file -- no live database needed
- if that is done, than schema discovery can also be used to validate a live database against a migration state
- separately the enums used to identify table names and columns are also code generated, since those in the migration files are actually how'd they may have existed in the past before the latest migration steps because things could have been dropped and/or renamed -- removes copy and pasting from the migration files to something referred to by `sea-query` invocations.

**Reply by tyt2y3** [reply]
+ there is a crate `sea-orm-migration` which depends on `sea-orm`. what do you expect `sea-migrations` to contain?
+ to clarify again, you don't need `sea-schema` for this step. All the code is you need is in `sea-orm-codegen`, which converts `sea-query` DDL statements into `Entity` representations
+ Yes, certainly
+ not sure what you mean about the last point

**Reply by kriswuollett** [reply]
We're getting more into my first impressions / opinions here, just to be clear :-) I'm really liking `sea-query`.

#### Migrations

The `sea-migrations` name and separation since having `orm` in the name could be confusing, since it really doesn't have to be used with the ORM.

#### Entity generation

For current state, all I see is the entity transformer working on create table statements, so I'd roughly have to reduce a list of create, update, and delete statements into just table create statements:

https://github.com/SeaQL/sea-orm/blob/edd6c9b3622ed22d77835e4b0de854fddf5b2f22/sea-orm-codegen/src/entity/transformer.rs#L12

I'd think things would be more capable if it was a a general schema struct that had things like tables, indices, etc., with the extra info on how relations are managed for the ORM system. In the bakery example below, that'd be saying it would be more worth the effort to transform `[CreateTable, AlterTable, AlterTable]` into a singular `Schema` instead of `[CreateTable]` to do offline entity generation.

#### Enums for naming

Let's say you aren't using the ORM at all, but just the migrations. And following along with something like in the [tutorial](https://github.com/SeaQL/sea-orm-tutorial/blob/0b2148772c55ba6a3b22fe195690188084cc2d32/tutorials-book/src/ch01-03-migration-api.md?plain=1#L104-L110):

```rust
#[derive(Iden)]
pub enum Bakery {
    Table,
    Id,
    Name,
    ProfitMargin,
}
```

Then I can look up the bakery with `sea-query` and reuse the enum for the naming:

```rust
        let (sql, arguments) = Query::select()
            .columns([
                Bakery::Name,
                Bakery::ProfitMargin,
            ])
            .from(Bakery::Table)
            .and_where(Expr::col(schema::Bakery::Id).eq(request.bakery_id.to_owned()))
            .build_sqlx(SqliteQueryBuilder);

        let row = sqlx::query_with(&sql, arguments)
            .fetch_one(&self.pool)
            .await
            .map_err(|err| match err {
                sqlx::Error::RowNotFound => Error::not_found(""),
                _ => {
                    event!(
                        Level::ERROR,
                        error = err.to_string(),
                        "Failed to lookup bakery in database"
                    );
                    Error::internal_error("")
                }
            })?;
```

However if I need to replace `ProfitMargin` with `Revenue` and `Expenses`, then the second migration will have this enum:

```rust
#[derive(Iden)]
pub enum Bakery {
    Table,
    Id,
    Name,
    ProfitMargin,
    Revenue,  // nullable
    Expenses,  // nullable
}
```

I'd update the database, and then have a third migration with this enum:

```rust
#[derive(Iden)]
pub enum Bakery {
    Table,
    Id,
    Name,
    Revenue,  // not null
    Expenses,  // not null
}
```

Since the source code for the migrations shouldn't be changed after being applied, and my "bakery database repo" service won't depend on the migrations library, since there would be a separate CLI tool that ops would use, then the repo code needs its own copy of the enum that only has the fields it is allowed to use.

**Reply by tyt2y3** [reply]
+ currently `sea-orm-migration` have to depends on `sea-orm` because it uses the `DatabaseConnection` from `sea-orm`, which provides a nice abstraction over all SQLx backends. To make it possible, I guess we need to extract that into it's own library.
+ I see, then really, we currently have the facility to perform this reduction operation. `TableCreateStatement` actually includes `IndexCreateStatement`. I would argue, the interface should still be based around `TableCreateStatement`:
```rust
enum DDLStatement {
    TableCreateStatement,
    TableAlterStatement,
    IndexCreateStatement,
    IndexDropStatement,
    ...
}

fn reduce(operations: Vec<DDLStatement>) -> TableCreateStatement;
```
+ So do you mean the `Iden` enums for use with SeaQuery should also be generated, in case a crate does not depend on SeaORM? Because with auto-generated SeaORM entities, one can also use them with SeaQuery. (Entity::Column is a enum that implements sea_query::Iden)

**Comment by kriswuollett** [comment]
> * currently `sea-orm-migration` have to depends on `sea-orm` because it uses the `DatabaseConnection` from `sea-orm`, which provides a nice abstraction over all SQLx backends. To make it possible, I guess we need to extract that into it's own library.

BTW, I was unable to initially figure out how to make my service database-agnostic, but now that you mention this, I realized the project already has an [example in `sea-query`](https://github.com/SeaQL/sea-query/blob/8ce9930b075aaa204a1b2adf08b234240e66fb57/examples/sqlx_any/src/main.rs). That's a really nice feature that probably deserves a mention in the [main readme](https://github.com/SeaQL/sea-query/blob/8ce9930b075aaa204a1b2adf08b234240e66fb57/README.md). Will give that approach when I need to support an additional database type.

Perhaps it is just a marketing thing. I'd just think having it a separate crate can help pull developers to the project. I only came across SeaQL since I originally wanted to try an ORM, and `sqlx` mentions [SeaQL as one](https://github.com/launchbadge/sqlx/blob/d0fbe7feff4b3a200cf0453417d5e53bd011643a/README.md?plain=1#L222). Some developers may like generating migrations with code instead of raw SQL files in a directory. 

The last database migration library I used was [`alembic`](https://alembic.sqlalchemy.org/en/latest/tutorial.html#create-a-migration-script) which I'm influenced by. Which reminds me about another offline feature that locked-down projects would like: offline SQL generated for migrations. It seems like migration can only be done on a live database? The use case code-reviewing a SQL "patch" before handing it off to an admin or some other release process. SQL generated can include any sized slice of the migration revisions. Perhaps a possibility if looking to add things to the project roadmap.

> * So do you mean the `Iden` enums for use with SeaQuery should also be generated, in case a crate does not depend on SeaORM? Because with auto-generated SeaORM entities, one can also use them with SeaQuery. (Entity::Column is a enum that implements sea_query::Iden)

Yes, although thinking about it more, it's just now that I'm finding it preferable to copy and paste the `Iden` enums than use a live database to generate SeaORM entities. Whatever method, I'd use them to write raw queries, mention field names in error messages, or use with`sea-query`.

Well... as you can see from everything above, I try to do everything offline and as reproducible as possible. Offline support would probably be a better title for the discussion. :-)

**Reply by tyt2y3** [reply]
+ Just to make a note: SQLx Any has limitations. It's kind of the 'lowest common denominator' of all backends. For example, `chrono::DateTime` type mappings are not supported. The `DatabaseConnection` in `sea-orm` is enum based, so have access to all features of a particular backend. There had been talks of extracting it into its own crate. But I hope we wouldn't encounter any orphan rule limitation while doing so.

+ I agree with you. Who wants `NVARCHAR(255)` ? And the migration system also allows some programmability, for example to seed data. Or `ADD COLUMN IF NOT EXIST`.

+ Interesting. I think you can run them against a `MockDatabase`. e.g. https://github.com/SeaQL/sea-orm/blob/edd6c9b3622ed22d77835e4b0de854fddf5b2f22/src/database/mock.rs#L841-L848
I think that would do the trick

+ If you are able to contribute we can starting adding the `reduce()` feature into sea-schema. This is something that would benefit from more testing.

**Comment by kriswuollett** [comment]
> * If you are able to contribute we can starting adding the `reduce()` feature into sea-schema. This is something that would benefit from more testing.

I'm currently using SeaQL for the first time with one of the services I'm building. It's possible I could get into contributing, but I'd like to roll some things out on my end first so I'd have a holistic view of what to experiment/test end-to-end here, e.g., potential end goal of offline generation of migration sql over any revision range. I'll post over here in this or a new discussion when I think I'd be ready.

**Reply by tyt2y3** [reply]
Sure, no problem!
